### DISCUSSION TITLE
Retrieving tokens created by Firebase Auth

### DISCUSSION THREAD

**Discussion by Hannnes1** [Budibase/budibase#10289] [Help]
Hi,
We are exploring how we can use Budibase together with Firebase Auth. For the sign in part, we have created a custom component that signs in the user using the steps described in [the Firebase documentation](https://firebase.google.com/docs/auth/web/password-auth#sign_in_a_user_with_an_email_address_and_password). (We will add more sign in methods later).

So signing in works well. The problem comes when we want to do a call to our back-end (that is the only data source we have). We have to pass in the id token created by Firebase Auth, and that token has to be valid (has to refresh if it has expired).  Normally this is done with `firebase.auth().currentUser.getIdToken()`. Is this possible in Budibase? The code to get the token has to run on the client side.

Is there a solution to our issue, or do we need to change our approach entirely?

**Comment by melohagan** [comment]
Hey @Hannnes1 

I'm not familiar with the Firebase auth flow, so it may be worth having a look at the integration to see what may be missing?
https://github.com/Budibase/budibase/blob/develop/packages/server/src/integrations/firebase.ts

It's also worth noting that we use the [@google-cloud/firestore](https://github.com/googleapis/nodejs-firestore) library, which from memory had limitations compared to some of the other libraries. I can't remember the specifics of the differences, but for the Firestore connector we expect a private key. 
I think a new different library has to be used to support Firebase authentication but I'm not certain.

If it helps, you can pass binding values from the front-end through to a back-end custom query as outlined here: https://docs.budibase.com/docs/data-sources#using-query-bindings

**Reply by Hannnes1** [reply]
Hi,
Thanks for the reply.

Yes, [@google-cloud/firestore](https://github.com/googleapis/nodejs-firestore) is only for the Firestore database. We would need the [firebase](https://github.com/firebase/firebase-js-sdk) library for this to work.

A binding might work, but according to the documentation "JavaScript within Budibase is server-side synchronous only", and we need to run asynchronous client side JavaScript. But I guess that isn't possible then?

We could also save the token as a string when the user signs in, that way we would be able to pass it in a binding. The issue then is that the token expires after 30 minutes, so we would have to find a way to call `firebase.auth().currentUser.getIdToken()` (which refreshes the token) when that happens.
